coreneuron/network/multisend_setup.cpp

```diff
@@ -451,151 +406,130 @@ static int setup_target_lists(int use_phase2, int** r_return) {
     // gid%nhost rank. The intermediate rank can then construct the
     // list of target ranks for the gids it gets.
 
-    // scnt is number of input gids from target
-    scnt = newintval(0, nhost);
-    for (Gid2IPS::iterator gid2in_it = gid2in.begin(); gid2in_it != gid2in.end(); ++gid2in_it) {
-        int gid = gid2in_it->first;
-        ++scnt[gid % nhost];
-    }
+    {
+        // scnt is number of input gids from target
+        std::vector<int> scnt1(nhost, 0);
+        for (const auto& g: gid2in) {
+            int gid = g.first;
+            ++scnt1[gid % nhost];
+        }
 
-    // s are the input gids from target to be sent to the various intermediates
-    sdispl = newoffset(scnt, nhost);
-    s = newintval(0, sdispl[nhost]);
-    for (Gid2IPS::iterator gid2in_it = gid2in.begin(); gid2in_it != gid2in.end(); ++gid2in_it) {
-        int gid = gid2in_it->first;
-        s[sdispl[gid % nhost]++] = gid;
-    }
-    // Restore sdispl for the message.
-    del(sdispl);
-    sdispl = newoffset(scnt, nhost);
+        // s are the input gids from target to be sent to the various intermediates
+        const std::vector<int> sdispl1 = newoffset(scnt1);
+        // Make an usable copy
+        auto sdispl1_ = sdispl1;
+        std::vector<int> s1(sdispl1[nhost], 0);
+        for (const auto& g: gid2in) {
+            int gid = g.first;
+            s1[sdispl1_[gid % nhost]++] = gid;
+        }
 
-    all2allv_int(s, scnt, sdispl, r, rcnt, rdispl, "gidin to intermediate");
-    del(s);
-    del(scnt);
-    del(sdispl);
-    // r is the gids received by this intermediate rank from all other ranks.
+        std::vector<int> r1;
+        std::vector<int> rdispl1;
+        std::tie(r1, rdispl1) = all2allv_int(s1, scnt1, sdispl1, "gidin to intermediate");
+        // r1 is the gids received by this intermediate rank from all other ranks.
 
-    // Construct hash table for finding the target rank list for a given gid.
-    gid2tarlist = new Int2TarList;
-    // Now figure out the size of the target list for each distinct gid in r.
-    for (int i = 0; i < rdispl[nhost]; ++i) {
-        TarList* tl;
-        Int2TarList::iterator itl_it = gid2tarlist->find(r[i]);
-        if (itl_it != gid2tarlist->end()) {
-            tl = itl_it->second;
-            tl->size += 1;
-        } else {
-            tl = new TarList();
-            tl->size = 1;
-            (*gid2tarlist)[r[i]] = tl;
+        // Now figure out the size of the target list for each distinct gid in r1.
+        for (const auto& gid: r1) {
+            auto tar = gid2tarlist[gid];
+            ++(tar->size);
         }
```

Comment:
rest of the code is same but seems like some logic is changed for above section? could you elaborate?

Reply:
this is the place of the segfault, so let's remember what I did here